modules/RTC/ortc/RTCPeerConnection.js

```diff
@@ -0,0 +1,532 @@
+
+// Whether we want to mimic the new Promises based RTCPeerConnection or use the
+// old callbacks based one.
+const USE_PROMISES = false;
+
+/**
+ * RTCPeerConnection shim for ORTC based endpoints (such as Edge).
+ *
+ * The interface is based on the W3C specification of 2015, which matches
+ * the implementation of Chrome nowadays:
+ *
+ *   https://www.w3.org/TR/2015/WD-webrtc-20150210/
+ */
+export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+    /**
+     */
+    constructor(pcConfig) {
+        super();
+
+        logger.debug('constructor() pcConfig:', pcConfig);
+
+        // Closed flag.
+        this._closed = false;
+
+        // Local RTCSessionDescription.
+        this._localDescription = null;
+
+        // Remote RTCSessionDescription.
+        this._remoteDescription = null;
+
+        // Set of local MediaStreams.
+        this._localStreams = new Set();
+
+        // Set of remote MediaStreams.
+        this._remoteStreams = new Set();
+
+        // RTCPeerConnection signalingState.
+        this._signalingState = 'stable';
+
+        // RTCPeerConnection iceGatheringState.
+        // NOTE: This should not be needed, but Edge does not implement
+        // iceGatherer.state.
+        this._iceGatheringState = 'new';
+
+        // Create a RTCIceGatherer.
+        this._iceGatherer = this._createIceGatherer(pcConfig);
+
+        // Create a RTCIceTransport.
+        this._iceTransport = this._createIceTransport(this._iceGatherer);
+    }
+
+    /**
+     */
+    get signalingState() {
+        return this._signalingState;
+    }
+
+    /**
+     */
+    get iceGatheringState() {
+        return this._iceGatheringState;
+    }
+
+    /**
+     */
+    get iceConnectionState() {
+        return this._iceTransport.state;
+    }
+
+    /**
+     */
+    get localDescription() {
+        return this._localDescription;
+    }
+
+    /**
+     */
+    get remoteDescription() {
+        return this._remoteDescription;
+    }
+
+    /**
+     */
+    close() {
+        if (this._closed) {
+            return;
+        }
+
+        this._closed = true;
+
+        logger.debug('close()');
+
+        this._updateAndEmitSignalingStateChange('closed');
+
+        // Close iceGatherer.
+        // NOTE: Not implemented by Edge.
+        try {
+            this._iceGatherer.close();
+        } catch (error) {
+            logger.warn(`iceGatherer.close() failed:${error}`);
+        }
+
+        // Close iceTransport.
+        try {
+            this._iceTransport.stop();
+        } catch (error) {
+            logger.warn(`iceTransport.stop() failed:${error}`);
+        }
+
+        // Clear local/remote streams.
+        this._localStreams.clear();
+        this._remoteStreams.clear();
+
+        // TODO: Close and emit more stuff.
+    }
+
+    /**
+     */
+    createOffer(options, callback, errback) {
+        logger.debug('createOffer() options:', options);
+
+        if (USE_PROMISES) {
```

Comment:
You are going to hate me... but I'd like to be able to use promises or callbacks without having to change a global.